form2request/_base.py

```diff
@@ -0,0 +1,220 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Any, Dict, Iterable, Optional, Tuple, Union, cast
+from urllib.parse import urlencode, urljoin, urlsplit, urlunsplit
+
+from w3lib.html import strip_html5_whitespace
+
+if TYPE_CHECKING:
+    from lxml.html import FormElement  # nosec
+    from lxml.html import HtmlElement  # nosec
+
+FormdataVType = Union[str, Iterable[str]]
+FormdataKVType = Tuple[str, FormdataVType]
+FormdataType = Optional[Union[Dict[str, FormdataVType], Iterable[FormdataKVType]]]
+
+
+def _is_listlike(x: Any) -> bool:
+    """Return ``True`` if *x* is a list-like object or ``False`` otherwise.
+
+    A list-like object is an iterable, excluding strings or bytes.
+    """
+    return hasattr(x, "__iter__") and not isinstance(x, (str, bytes))
+
+
+def _url(form: FormElement, click_element: HtmlElement | None) -> str:
+    if form.base_url is None:
+        raise ValueError(f"{form} has no base_url set.")
+    action = (
+        click_element.get("formaction") if click_element is not None else None
+    ) or form.get("action")
+    if action is None:
+        return form.base_url
+    return urljoin(form.base_url, strip_html5_whitespace(action))
+
+
+USER = object()
+
+
+def _method(
+    form: FormElement, click_element: HtmlElement | None, method: None | str
+) -> str:
+    if method:
+        method_src = USER
+    else:
+        if click_element is not None:
+            method = click_element.get("formmethod")
+        if method:
+            method_src = click_element
+        else:
+            method = form.method
+            assert method is not None  # lxml’s form.method is always filled
+            method_src = form
+    method = method.upper()
+    if method not in {"GET", "POST"}:
+        if method_src is USER:
+            raise ValueError(
+                f"The specified form method ({method!r}) is not supported."
+            )
+        if method_src is click_element:
+            raise NotImplementedError(
+                f"Found unsupported form method {method!r} in the formmethod "
+                f"attribute of the submission button."
```

Comment:
What is the browser behavior when an unsupported method is used? I think it may be good to follow it in the library. For example, if browsers fall back to GET or POST, I'd also fall back to GET or POST instead of raising an exception. There could be an optional strict mode.

Reply:
You are right, they fall back to GET for invalid values. Will do that.

I think we should still raise an exception for `dialog`, though, as that value causes the form not to be convertible into a request. But maybe `NotImplementedError` is not the right exception, maybe it should be a `ValueError`.